src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -0,0 +1,385 @@
+                                    metric.value() != raft::distance::DistanceType::L2Expanded
+                                    && metric.value() != raft::distance::DistanceType::L2Unexpanded
+                                    && metric.value() != raft::distance::DistanceType::InnerProduct
+                               ) {
+                                LOG_KNOWHERE_WARNING_ << 
+                                    "selected metric not supported in RAFT IVF indexes: " << 
+                                    ivf_raft_cfg.metric_type;
+                                return Status::invalid_metric_type;
+                            }
+
+                            auto scoped_device = detail::device_setter{
+                                *ivf_raft_cfg.gpu_ids.begin()
+                            };
+                            res_ = std::make_unique<raft::device_resources>();
+                            auto rows = dataset.GetRows();
+                            auto dim = dataset.GetDim();
+                            auto* data = reinterpret_cast<float const*>(dataset.GetTensor());
+
+                            auto stream = res_->get_stream();
+                            auto data_gpu = rmm::device_uvector<float>(rows * dim, stream);
+                            RAFT_CUDA_TRY(
+                                    cudaMemcpyAsync(
+                                        data_gpu.data(),
+                                        data,
+                                        data_gpu.size() * sizeof(float),
+                                        cudaMemcpyDefault,
+                                        stream.value()
+                                        )
+                                    );
+                            if constexpr (std::is_same_v<detail::raft_ivf_flat_index, T>) {
+                                auto build_params = raft::neighbors::ivf_flat::index_params{};
+                                build_params.metric = metric.value();
+                                build_params.n_lists = ivf_raft_cfg.nlist;
+                                gpu_index_ = raft::neighbors::ivf_flat::build<
+                                    float, std::int64_t
+                                    >(
+                                            *res_,
+                                            build_params,
+                                            data_gpu.data(),
+                                            rows,
+                                            dim
+                                     );
+                            } else if constexpr (std::is_same_v<detail::raft_ivf_pq_index, T>) {
+                                auto build_params = raft::neighbors::ivf_pq::index_params{};
+                                build_params.metric = metric.value();
+                                build_params.n_lists = ivf_raft_cfg.nlist;
+                                build_params.pq_bits = ivf_raft_cfg.nbits;
```

Comment:
Should probably expose pq_dim here as well, in addition to the coarse quantizer iterations. If it's possible to do in Knowhere, we should also consider smaller precision sizes for the training/query datasets as well.

Reply:
Agreed! Same thoughts on above in terms of possibly doing this in a follow-up or getting help from other contributors.